### PR TITLE
Turn to an angle based on vision

### PR DESCRIPTION
**Pull request by having11**

### CODE REVIEW

File: src/main/cpp/RobotContainer.cpp [hunk 1]
```diff
@@ -200,6 +187,17 @@ void RobotContainer::ConfigureButtonBindings() {
           .AndThen(m_leds.BlinkingFace())
           .AndThen(m_leds.Idling()));
 
+  m_driverController
+      .POVUp(frc2::CommandScheduler::GetInstance().GetActiveButtonLoop())
+      .CastTo<frc2::Trigger>()
+      .OnTrue(m_rightClimb.MoveToPositionAbsolute(20_in).AlongWith(
```
Comment (having11) [inline src/main/cpp/RobotContainer.cpp:193]: Make these heights constants

File: src/main/cpp/RobotContainer.cpp [hunk 2]
```diff
@@ -396,6 +398,32 @@ void RobotContainer::Periodic() {
         inRange = m_aimbotEnabled && true;
         m_turnToPose.SetTargetPose(location.trackedPose);
 
+        if (inRange && m_autoScoringEnabled && location.scoringRadius &&
```
Comment (having11) [inline src/main/cpp/RobotContainer.cpp:401]: Icky

File: src/main/cpp/RobotContainer.cpp [hunk 3]
```diff
@@ -405,24 +433,41 @@ void RobotContainer::Periodic() {
...
-      m_turnToPose.SetTargetPose(targetPose.value());
+    if (bestTarget) {
+      m_shouldAim = m_aimbotEnabled;
+      auto centerDiff = bestTarget.value().centerX;
 
-      frc::SmartDashboard::PutNumber(
-          "TURN TO POSE TARGET deg",
-          targetPose.value().Rotation().Degrees().value());
+      auto normalizedTargetAngle = TurnToPose::NormalizeScalar(
+          centerDiff.value(), -30.0, 30.0, -1.0, 1.0);
```
Comment (having11) [inline src/main/cpp/RobotContainer.cpp:456]: Constants

File: src/main/cpp/subsystems/ScoringSubsystem.cpp [hunk 4]
```diff
@@ -95,9 +103,13 @@ void ScoringSubsystem::SpinAmp(double upperPercentage, double lowerPercentage) {
   frc::SmartDashboard::PutNumber("Amp lower target %", lowerPercentage);
 }
 
-void ScoringSubsystem::SpinSpeaker(double upperPercentage, double lowerPercentage) {
-  m_speakerLowerSpinnyBoi.Set(-lowerPercentage);
-  m_speakerUpperSpinnyBoi.Set(-upperPercentage);
+void ScoringSubsystem::SpinSpeaker(double upperPercentage,
+                                   double lowerPercentage) {
+  // m_speakerLowerSpinnyBoi.Set(-lowerPercentage);
```
Comment (having11) [inline src/main/cpp/subsystems/ScoringSubsystem.cpp:108]: Should this be a macro/flag to use either `Set` or `RunWithVelocity`? Would be the same across all `Spin...` methods.

File: src/main/cpp/utils/TurnToPose.cpp [hunk 5]
```diff
@@ -54,6 +63,26 @@ units::degree_t TurnToPose::GetAngleFromOtherPose(
...
   m_startPose = m_poseGetter();
   m_targetPose = pose;
+  m_targetAngle = std::nullopt;
+}
+
+void TurnToPose::SetTargetAngleRelative(units::degree_t angle) {
+  m_startPose = m_poseGetter();
+  auto wpiTransformation = frc::Transform2d(0_m, 0_m, angle);
+  auto wpiFinalPose = m_startPose.TransformBy(wpiTransformation);
+  m_targetAngle = wpiFinalPose.Rotation().Degrees();
```
Comment (having11) [inline src/main/cpp/utils/TurnToPose.cpp:73]: Do we need this since the line below works fine?

File: src/main/include/RobotContainer.h [hunk 6]
```diff
@@ -154,10 +161,14 @@ class RobotContainer {
   frc2::CommandPtr IntakeTarget();
 
   frc2::CommandPtr autoCommand = frc2::InstantCommand([] {}).ToPtr();
+  frc2::CommandPtr autoScoreCommand = frc2::InstantCommand([] {}).ToPtr();
 
   bool m_aimbotEnabled = false;
```
Comment (having11) [inline src/main/include/RobotContainer.h:166]: Icky

File: src/main/include/subsystems/DriveSubsystem.h [hunk 7]
```diff
@@ -207,7 +207,7 @@ class DriveSubsystem : public frc2::SubsystemBase {
        m_frontRight.GetPosition(), m_rearRight.GetPosition()},
       frc::Pose2d{0_m, 0_m, 0_rad},
       {0.1, 0.1, 0.1},
-      {1.0, 1.0, 1.0}};
+      {0.9, 0.9, 0.9}};
```
Comment (having11) [inline src/main/include/subsystems/DriveSubsystem.h:210]: Constants?

File: src/main/include/utils/Commands/IntakeCommands.h [hunk 8]
```diff
@@ -68,46 +69,31 @@ static frc2::CommandPtr DowntakeUntilPresent(IntakeSubsystem* intake,
...
-             },
-             // onEnd
-             [intake, scoring](bool interupted) {},
-             // isFinished
-             [intake, scoring] { return intake->NotePresentUpperAll(); },
-             // req
-             {intake, scoring})
-      .ToPtr()
+  return frc2::InstantCommand([intake, scoring] {
+                 scoring->SpinAmp(0.1, 0.1);
```
Comment (having11) [inline src/main/include/utils/Commands/IntakeCommands.h:73]: Constants

File: src/main/include/utils/ModifiableChooser.h [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+  std::map<std::string, T> m_map;
+
+  std::string m_defaultChoice = "";
+  int m_instance;
+  std::string m_previousValue;
+  std::function<void(std::optional<T>)> m_listener;
+  std::atomic_int m_instances{0};
+  std::string m_selected;
+  std::recursive_mutex m_mutex;
```
Comment (having11) [inline src/main/include/utils/ModifiableChooser.h:38]: Mutexes are neat

File: src/main/include/utils/TurnToPose.h [hunk 10]
```diff
@@ -46,23 +51,32 @@ class TurnToPose {
...
     return m_targetPose;
   }
 
+  inline std::optional<units::degree_t> GetTargetAngle() const {
+    return m_targetAngle;
+  }
+
   inline units::degree_t GetTargetHeading() const { return m_targetHeading; }
 
+  static double NormalizeScalar(double x, double from_min, double from_max, double to_min, double to_max) {
```
Comment (having11) [inline src/main/include/utils/TurnToPose.h:68]: Could declare as `inline`

**Review by nobody5050 — APPROVED** [review]

**Pull request merged**